codex/blockexchange/network/network.nim

```diff
@@ -34,19 +34,19 @@ const
   MaxInflight* = 100
 
 type
-  WantListHandler* = proc(peer: PeerId, wantList: WantList): Future[void] {.gcsafe.}
-  BlocksDeliveryHandler* = proc(peer: PeerId, blocks: seq[BlockDelivery]): Future[void] {.gcsafe.}
-  BlockPresenceHandler* = proc(peer: PeerId, precense: seq[BlockPresence]): Future[void] {.gcsafe.}
-  AccountHandler* = proc(peer: PeerId, account: Account): Future[void] {.gcsafe.}
-  PaymentHandler* = proc(peer: PeerId, payment: SignedState): Future[void] {.gcsafe.}
+  WantListHandler* = proc(peer: PeerId, wantList: WantList): Future[void] {.gcsafe, raises: [].}
```

Comment:
This will necessitate deeper changes, as many of those callbacks are actually not implemented as async procs but as [wrappers that do things and then return a future from a delegate](https://github.com/codex-storage/nim-codex/blob/master/codex/blockexchange/network/network.nim#L117).

Reply:
the recommendation for such wrappers in v4 is to mark them `{.async: (raw: true).}` - again, not necessary.

Reply:
oh, and the wrappers would remain compatible even if you annotated with `{.async.}` here (`{.async.}` in that case applies only to the forward declaration)

Reply:
Oh, thanks for pointing that out. I had in my head that a plain `async` proc would have to return `Future[void].Raising([CatchableError])` which would not be compatible with something that returns plain `Future`, but I now realize that `InternalRaisesFuture[T, E]* = ref object of Future[T]` so it all fits in quite nicely. :-)

Reply:
Oh, thanks for pointing that out. Why does that work, though? I had it in my head that a plain `async proc` would return `InternalRaisesFuture[void, (CatchableError,)]`, which would not be compatible with something that returns a plain `Future[void]`. But indeed this compiles:
```nim
type Callback = proc(): Future[void] {.async.}

proc returnsRawFuture: Future[void] = newFuture[void]("bla")

proc caller(cb: Callback) {.async.} = 
  await cb()

waitFor caller(returnsRawFuture)
  ```
even if this doesn't:
```nim
var a: Future[void].Raising([CatchableError])
var b: Future[void]

a = b
```

Reply:
no raises in async -> plain future for backwards compat

Reply:
I've:

* annotated the above forward decls with `async`;
* annotated the actual callbacks with `async: (raw: true)` where applicable. 

I confess I did not track down every `proc` that returns a `Future` and is not annotated with `async` across the entire codebase, I suppose can do that migration incrementally as I bump into them. If we're alright with this, then I will resolve this conversation.